File: src/GameState.zig
//TODO UPDATE LAST COMMAND ON FRAME!!!!

const std = @import("std");
const GameState = @This();
const bwapi = @import("bwapi_module.zig");
const bw = @import("bwenums.zig");
const events = @import("events.zig");
const BWAPIError = error{
    UnitNotFound,
    UnitRecordNotFound
};

//members
unit_list: std.AutoHashMap(c_int, UnitRecord),
self_player_id: c_int,
self_race: c_int,
enemy_race: c_int,

//definitions
const UnitRole = enum { 
    MINER, 
    GAS_GATHER, 
    BUILDER, 
    SCOUT, 
    ARMY, 
    UNKNOWN 
};

//weighted random selection
const Strategy = enum{
    FOUR_POOL,
    OGRE_ZERG
};

const UnitRecord = struct { 
    id: c_int, 
    type: bw.UnitType, 
    is_friendly: bool,
    is_visible: bool,
    role: UnitRole, 
    role_verified: bool,
    last_command: bw.UnitCommandType
};

pub fn init(self: *GameState, allocator: std.mem.Allocator, Broodwar: ?*bwapi.Game) void {
    //allocations
    self.unit_list = std.AutoHashMap(c_int, UnitRecord).init(allocator);
    //determine races
    const me = bwapi.Game_self(Broodwar);
    const enemy = bwapi.Game_enemy(Broodwar);
    self.self_race =  bwapi.Player_getRace(me).id;
    self.enemy_race = bwapi.Player_getRace(enemy).id;
}

pub fn deinit(self: *GameState) void {
    self.unit_list.deinit();
}

//not handling nuke detect or system messages
pub fn updateGameStateFromEvents(self: *GameState, 
    new_events: std.ArrayList(events.UnitEvent), Broodwar: ?*bwapi.Game) void{
    const EventType = events.EventType;
    for (new_events.items) |event| {
        switch (event.type) {
            //have to handle unit registration failure here so not passed up to toframe
            EventType.UnitCreate => registerNewUnit(self, event, Broodwar) catch |err|{
                if (err == BWAPIError.UnitNotFound) std.debug.print("UnitNotFound error on registration\n", .{});
            },
            EventType.UnitDiscover => registerNewUnit(self, event, Broodwar) catch |err|{
                if (err == BWAPIError.UnitNotFound) std.debug.print("UnitNotFound error on registration\n", .{});
            },
            EventType.UnitMorph => updateUnitType(self, event, Broodwar) catch |err|{
                if (err == BWAPIError.UnitNotFound) std.debug.print("UnitNotFound error on morph\n", .{});
                if (err == BWAPIError.UnitRecordNotFound) std.debug.print("UnitRecordNotFound on morph\n", .{});
            },
            EventType.UnitRenegade => updateUnitType(self, event, Broodwar) catch |err|{
                if (err == BWAPIError.UnitNotFound) std.debug.print("UnitNotFound error on morph\n", .{});
                if (err == BWAPIError.UnitRecordNotFound) std.debug.print("UnitRecordNotFound on morph\n", .{});
            },
            EventType.UnitDestroy => rmUnit(self, event) catch |err|{
                if (err == BWAPIError.UnitRecordNotFound) std.debug.print("UnRecordNotFound on remove\n", .{});
            },
            EventType.UnitHide =>{
                const record = self.unit_list.getPtr(event.unit_id) orelse return std.debug.print("UnitRecordNotFound on hide\n", .{});
                record.is_visible = false;
            },
            EventType.UnitShow =>{
                const record = self.unit_list.getPtr(event.unit_id) orelse return std.debug.print("UnitRecordNotFound on show\n", .{});
                record.is_visible = true;
            },
            else => continue,
        }
    }
}

//called once we are sure a unit is new to us
fn registerNewUnit(self: *GameState, event: events.UnitEvent, Broodwar: ?*bwapi.Game) !void{
    //check if unit is found by id
    const unit_ptr = bwapi.Game_getUnit(Broodwar, event.unit_id);
    if (unit_ptr) |u_ptr_val|{
        const unit_owner = bwapi.Unit_getPlayer(u_ptr_val) orelse unreachable; //hard crash
        //create and append record
        const new_unit: UnitRecord = .{
            .id = event.unit_id,
            .type = @enumFromInt(bwapi.Unit_getType(unit_ptr).id),
            .is_visible = true,
            .is_friendly = (bwapi.Player_getID(unit_owner) == self.self_player_id),
            .role = UnitRole.UNKNOWN,
            .role_verified = false,
        };
        try self.unit_list.put(new_unit.id, new_unit);
    }else{
        //handle null unit ptr
        return BWAPIError.UnitNotFound;
    }
}

//only applicable for morph? updates type on unit per event
fn updateUnitType(self: *GameState, event: events.UnitEvent, Broodwar: ?*bwapi.Game) !void{
    //check if unit is found by id
    const unit_ptr = bwapi.Game_getUnit(Broodwar, event.unit_id);
    if (unit_ptr)|u_ptr_val|{
        //we have to check ownership as well as type specifically for geyser to extractor case
        const unit_owner = bwapi.Unit_getPlayer(u_ptr_val) orelse unreachable; //hard crash
        const record = self.unit_list.getPtr(event.unit_id) orelse return BWAPIError.UnitRecordNotFound;
        record.type = @enumFromInt(bwapi.Unit_getType(u_ptr_val).id);
        record.is_friendly = (bwapi.Player_getID(unit_owner) == self.self_player_id);
    }else{
        return BWAPIError.UnitNotFound;
    }
}

fn rmUnit(self: *GameState, event: events.UnitEvent) !void{
    //check if unit is found by id and remove
    const removed = self.unit_list.remove(event.unit_id);
    if (!removed) return BWAPIError.UnitRecordNotFound;
}

File: src/gamestate.c
#include "gamestate.h"

#include "Player.h"
#include "Unit.h"

typedef enum BwapiError { BW_OK = 0, BW_UNIT_NOT_FOUND, BW_UNIT_RECORD_NOT_FOUND } BwapiError;

void gamestate_init(GameState* gs, Game* broodwar)
{
    intmap_init(&gs->unit_list, sizeof(UnitRecord));
    gs->directive_list = (DirectiveVec){0};
    gs->task_list = (TaskVec){0};

    Player* me = Game_self(broodwar);
    Player* enemy = Game_enemy(broodwar);
    gs->self_player_id = Player_getID(me);
    gs->self_race = Player_getRace(me).id;
    gs->enemy_race = enemy ? Player_getRace(enemy).id : RACE_Unknown;
}

void gamestate_deinit(GameState* gs)
{
    intmap_free(&gs->unit_list);
    vec_free(&gs->directive_list);
    vec_free(&gs->task_list);
}

UnitRecord* gamestate_find_unit(GameState* gs, int unit_id)
{
    return intmap_get(&gs->unit_list, unit_id);
}

/* called once we are sure a unit is new to us */
static BwapiError register_new_unit(GameState* gs, UnitEvent ev, Game* broodwar)
{
    Unit* unit = Game_getUnit(broodwar, ev.unit_id);
    if (!unit) return BW_UNIT_NOT_FOUND;
    Player* owner = Unit_getPlayer(unit);

    UnitRecord rec = {
        .id = ev.unit_id,
        .type = (UnitTypeId)Unit_getType(unit).id,
        .is_friendly = owner && Player_getID(owner) == gs->self_player_id,
        .is_visible = true,
        .role = ROLE_UNKNOWN,
        .role_verified = false,
    };
    if (intmap_put(&gs->unit_list, rec.id, &rec)) {
        LOG("out of memory registering unit %d\n", rec.id);
    }
    return BW_OK;
}

/* updates type and ownership on morph/renegade (geyser -> extractor changes both) */
static BwapiError update_unit_type(GameState* gs, UnitEvent ev, Game* broodwar)
{
    Unit* unit = Game_getUnit(broodwar, ev.unit_id);
    if (!unit) return BW_UNIT_NOT_FOUND;
    UnitRecord* rec = gamestate_find_unit(gs, ev.unit_id);
    if (!rec) return BW_UNIT_RECORD_NOT_FOUND;
    Player* owner = Unit_getPlayer(unit);
    rec->type = (UnitTypeId)Unit_getType(unit).id;
    rec->is_friendly = owner && Player_getID(owner) == gs->self_player_id;
    return BW_OK;
}

static BwapiError rm_unit(GameState* gs, UnitEvent ev)
{
    return intmap_remove(&gs->unit_list, ev.unit_id) ? BW_OK : BW_UNIT_RECORD_NOT_FOUND;
}

static void set_visible(GameState* gs, UnitEvent ev, bool visible, const char* what)
{
    UnitRecord* rec = gamestate_find_unit(gs, ev.unit_id);
    if (!rec) { LOG("UnitRecordNotFound on %s (id %d)\n", what, ev.unit_id); return; }
    rec->is_visible = visible;
}

/* not handling nuke detect or system messages */
void gamestate_update_from_events(GameState* gs, const UnitEventVec* events, Game* broodwar)
{
    for (size_t i = 0; i < events->len; i++) {
        UnitEvent ev = events->items[i];
        BwapiError err;
        switch (ev.type) {
        case EV_UnitCreate:
        case EV_UnitDiscover:
            err = register_new_unit(gs, ev, broodwar);
            if (err == BW_UNIT_NOT_FOUND) LOG("UnitNotFound error on registration (id %d)\n", ev.unit_id);
            break;
        case EV_UnitMorph:
        case EV_UnitRenegade:
            err = update_unit_type(gs, ev, broodwar);
            if (err == BW_UNIT_NOT_FOUND) LOG("UnitNotFound error on morph (id %d)\n", ev.unit_id);
            if (err == BW_UNIT_RECORD_NOT_FOUND) LOG("UnitRecordNotFound on morph (id %d)\n", ev.unit_id);
            break;
        case EV_UnitDestroy:
            err = rm_unit(gs, ev);
            if (err == BW_UNIT_RECORD_NOT_FOUND) LOG("UnitRecordNotFound on remove (id %d)\n", ev.unit_id);
            break;
        case EV_UnitHide:
            set_visible(gs, ev, false, "hide");
            break;
        case EV_UnitShow:
            set_visible(gs, ev, true, "show");
            break;
        default:
            break;
        }
    }
}

int gamestate_get_active_directives(GameState* gs, DirectivePtrVec* out, Game* broodwar)
{
    Player* me = Game_self(broodwar);
    Race race = { gs->self_race };

    for (size_t i = 0; i < gs->directive_list.len; i++) {
        Directive* dir = &gs->directive_list.items[i];
        if (dir->status != STATUS_INACTIVE) continue;
        /* TODO logic to revisit certain done ones. rn done is done */

        bool satisfied = false;
        switch (dir->prereq.type) {
        case PREREQ_AI_FLAG:
            break; /* TODO flags not implemented */
        case PREREQ_SUPPLY:
            satisfied = Player_supplyUsed(me, race) == dir->prereq.qty;
            break;
        case PREREQ_TIMESTAMP:
            satisfied = Game_elapsedTime(broodwar) >= dir->prereq.qty;
            break;
        case PREREQ_UNIT_QTY:
            break; /* TODO count units of prereq.value in unit_list */
        default:
            break;
        }
        if (!satisfied) continue;

        if (vec_push(out, dir)) return -1;
        dir->status = STATUS_IN_PROGRESS;
    }
    return 0;
}

int gamestate_update_tasks_from_directives(GameState* gs, Game* broodwar)
{
    DirectivePtrVec new_dirs = {0};
    int rc = gamestate_get_active_directives(gs, &new_dirs, broodwar);
    if (rc == 0) {
        for (size_t i = 0; i < new_dirs.len; i++) {
            Directive* dir = new_dirs.items[i];
            /* TODO create tasks linked to this directive and append to gs->task_list */
            Game_printf(broodwar, "directive active: %s %d %s",
                        dir->command.type == CMD_BUILD ? "BUILD" : "CMD",
                        dir->command.target_qty, UnitTypeNames[dir->command.target_type]);
        }
    }
    vec_free(&new_dirs);
    return rc;
}
